src/Middleware/ResponseCaching/src/CacheEntry/CachedResponseBody.cs

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.ResponseCaching
+{
+    internal class CachedResponseBody
+    {
+        public CachedResponseBody(List<byte[]> segments, long length)
+        {
+            Segments = segments;
+            Length = length;
+        }
+
+        public List<byte[]> Segments { get; }
+
+        public long Length { get; }
+
+        public async Task CopyToAsync(PipeWriter destination, CancellationToken cancellationToken)
+        {
+            if (destination == null)
+            {
+                throw new ArgumentNullException(nameof(destination));
+            }
+
+            for (var segmentIndex = 0; segmentIndex < Segments.Count; segmentIndex++)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                var segmentLength = Segments[segmentIndex].Length;
+
+                var memory = destination.GetMemory(segmentLength);
+                Segments[segmentIndex].AsMemory().CopyTo(memory);
+
+                destination.Advance(segmentLength);
+
+                await destination.FlushAsync();
+            }
+
+            await destination.CompleteAsync();
```

Comment:
I wouldn't complete here, Kestrel will complete for you instead. I think this honestly may noop right now.

Reply:
Yeah I think it is not necessary given it has been flushed before.
I thought however this is the direction pipes are going to and so will be needed at some point, as there is will be guarantee that complete is not blocking even after flush.
Anyway as you said kestrel will do it anyway, so I don't think it makes a difference to do it early.